### ISSUE TITLE
Strange behavior in Node client

### ISSUE THREAD

**Issue by derekjn** [pipelinedb/pipelinedb#1213]
(From @EliSnow, originally part of #1204)

Here's a pared down js file which replicates the issue (let me know if you need more than this):
```js
"use strict";
let pg = require("pg");

let connectString = 'postgres://db_user:password@192.168.99.100/stats';
let delimiter = '\r\n';

pg.connect(connectString, (err, client, done) => {
  if (err) {
    console.error(err);
    return;
  }
  let data = '{"foo":"bar"}\r\n{"baz":"bux"}\r\n';
  client.query(
    {
      name: "junk",
      text: `INSERT INTO tsd_stream (js) SELECT unnest(string_to_array(left($1, -${delimiter.length}), $2)::jsonb[])`,
      values: [data, delimiter]
    },
    (err, response) => {
      done();
      if (err) {
        console.error(err);
        console.log(data);
      }
    });
});
```
This is using the npm library [ps](https://www.npmjs.com/package/pg) so a `npm install ps` will resolve that dependency. The stream `tsd_stream` has a single jsonb column `js`.

I'm actually experiencing two issues:
1) in the version above the error returned is `error: expected 1 prepared parameters but received 2`. 
2) my original version did not have `delimiter` as a parameter rather it was concatenated directly into the query string (since it is hardcoded into the server code, it is "safe"). However that resulted in bizarre parsing errors on each of the continuous views feeding off of `tsd_stream`. To demonstrate this problem if you have one or more continuous views using `tsd_stream` and you replace the above query object with:
```js
{
    name: "junk",
    text: `INSERT INTO tsd_stream (js) SELECT unnest(string_to_array(left($1, -${delimiter.length}), '${delimiter}')::jsonb[])`,
    values: [data]
}
```
In the database logs you will a json parsing error on every continuous view feeding off of `tsd_stream`. 
```
ERROR:  invalid input syntax for type json
DETAIL:  Expected end of input, but found "{".
```
If you change the `INSERT` into a `SELECT`, you will see the json is parsed correctly:
```js
{
    name: "junk",
    text: `SELECT unnest(string_to_array(left($1, -${delimiter.length}), '${delimiter}')::jsonb[])`,
    values: [data]
}
```
Which seems to indicate the error is happening when PipelineDB is parsing the query.

**Comment by EliSnow** [comment]
Let me know if you need help replicating. I'm running in Docker containers and I could put together a few Dockerfiles to demo the issues relatively easily.

**Comment by EliSnow** [comment]
I just realized there's a typo in my message in the first comment. That should be `npm install pg`.

**Comment by derekjn** [comment]
Thanks @EliSnow , I can repro this issue locally now, although I tweaked your script slightly because I was having some issues. I'm using the following:

```
var pg = require("pg");

var connectString = 'postgres://localhost/pipeline';
var delimiter = '\r\n';

pg.connect(connectString, function(err, client, done) {
  if (err) {
    console.error(err);
    return;
  }
  var data = '{"foo":"bar"}\r\n{"baz":"bux"}\r\n';
	var q = {
      name: "junk",
      text: 'INSERT INTO tsd_stream (js) SELECT unnest(string_to_array(left($1, -2), $2)::jsonb[])',
      //text: "SELECT unnest(string_to_array(left($1, -2), '\r\n')::jsonb[])",
      values: [data, delimiter]
	}
  client.query(q,
    function(err, result) {
      done();
      if (err) {
        console.error(err);
        console.log(data);
      }
    });
});
```

We'll be back here soon with any new information and a fix.

**Comment by EliSnow** [comment]
Thanks, @derekjn for looking into this. I've got another, possibly related test case for you to try.

First create this `CONTINUOUS VIEW`:
```sql
CREATE CONTINUOUS VIEW test AS
SELECT js FROM tsd_stream;
```
Then execute this script:
```js
var pg = require("pg");

var connectString = 'postgres://localhost/pipeline';
var delimiter = '\r\n';

var client = new pg.Client(connectString);
client.connect((err) => {
  if (err) {
    console.error(err);
    return;
  }
  var data = '[{"foo":"bar"},{"baz":"bux"}]';
  client.query(
    {
      name: "junk",
      text: 'INSERT INTO tsd_stream (js) SELECT jsonb_array_elements($1::jsonb)',
      values: [data]
    },
    (err, response) => {
      client.end();
      console.log(response.rows);
      if (err) {
        console.error(err);
        console.log(data);
      }
    });
});
```

You will notice only one jsonb object is shown in `test` when there should be two.

**Comment by derekjn** [comment]
Thanks. Also, a good sanity check is always to create a table that's identical to your stream, and run the client queries against the table instead of the stream. If the error persists, it is likely that the issue is with how the query is formed or possibly a bug in the client.

**Comment by derekjn** [comment]
So what's going on here is that originally only raw values were allowed to be inserted into streams. Because of this, we assume that there is a 1-1 mapping between parameters and stream columns. Since arbitrary expressions and queries can now be used in stream insertions, this is a bug. It's actually related to #958.

We'll have a fix out shortly.

**Comment by derekjn** [comment]
So it looks like the best fix this is going to be #958. Instead of pushing out a quick fix for #1213, we'd rather just spend the time to do this the right way. I've put #958 in release 0.8.2, which is slated to be released at the end of the month.

In the interim, the workaround is to simply not use parameters for anything other than columns (for streams, at least). So in your case you can just use `delimiter` to generate the prepared statement on the client if `delimiter` isn't changing. Alternatively, you can make `delimiter` an actual stream columns.

Hopefully that will work for you until the next release goes out.
